### ISSUE TITLE
F811 is not raised when it is needed

### ISSUE THREAD

**Issue by Denis-Alexeev** [astral-sh/ruff#11828]
Ruff does not raise F811 error when unused redefinition exists.

Keywords:
"F811", "redefinition"

Code snippet:
```python
"""module."""
class AAA:
    """ddd."""

    def world(self: "AAA") -> None:
        """Docstring."""

    hello = world

    def hello(self: "AAA") -> None:
        """Docstring."""

``` 
Command:
`ruff check --config /path/to/config /path/to/python/file.py`

ruff.toml:
```
target-version = "py312"

[lint]
select = ["ALL"]
```

ruff version:
ruff 0.4.8

**Comment by charliermarsh** [comment]
Not clear why it's not being raised here, but it should be AFAICT.

**Comment by ukyen8** [comment]
The shadowed binding can also be an Assignment. At the moment, Ruff checks shadowed binding for class, function, import, from import.

**Comment by ukyen8** [comment]
@charliermarsh I like using Ruff. I'd love to contribute to ruff and would like to work on this issue. Is it okay if I take this on?

**Comment by dhruvmanila** [comment]
@ukyen8 Yeah, go for it. The code for this rule logic is at https://github.com/astral-sh/ruff/blob/c53d55a483f0f81f0dcdc67f762bbf2c71e96773/crates/ruff_linter/src/checkers/ast/analyze/deferred_scopes.rs#L190-L309

 Don't hesitate to ask any questions if you're stuck either here or on Discord :)

**Comment by charliermarsh** [comment]
Fixed in https://github.com/astral-sh/ruff/pull/11961, forgot to close.
